cookbook/security/guard-authentication.rst

```diff
@@ -0,0 +1,577 @@
+    # {"message":"Username could not be found."}
+
+    # test with a working token
+    curl -H "X-AUTH-TOKEN: REAL" http://localhost:8000/
+    # the homepage controller is executed: the page loads normally
+
+Now, learn more about what each method does.
+
+.. _guard-auth-methods:
+
+The Guard Authenticator Methods
+-------------------------------
+
+Each authenticator needs the following methods:
+
+**getCredentials(Request $request)**
+    This will be called on *every* request and your job is to read the token (or
+    whatever your "authentication" information is) from the request and return it.
+    If you return ``null``, the rest of the authentication process is skipped. Otherwise,
+    ``getUser()`` will be called and the return value is passed as the first argument.
+
+**getUser($credentials, UserProviderInterface $userProvider)**
+    If ``getCredentials()`` returns a non-null value, then this method is called
+    and its return value is passed here as the ``$credentials`` argument. Your job
+    is to return an object that implements ``UserInterface``. If you do, then
+    ``checkCredentials()`` will be called. If you return ``null`` (or throw an
+    :ref:`AuthenticationException <guard-customize-error>`)
+    authentication will fail.
+
+**checkCredentials($credentials, UserInterface $user)**
+    If ``getUser()`` returns a User object, this method is called. Your job is to
+    verify if the credentials are correct. For a login for, this is where you would
+    check that the password is correct for the user. To pass authentication, return
+    ``true``. If you return *anything* else
+    (or throw an :ref:`AuthenticationException <guard-customize-error>`),
+    authentication will fail.
+
+**onAuthenticationSuccess(Request $request, TokenInterface $token, $providerKey)**
+    This is called after successful authentication and your job is to either
+    return a :class:`Symfony\\Component\\HttpFoundation\\Response` object
+    that will be sent to the client or ``null`` to continue the request
+    (e.g. allow the route/controller to be called like normal). Since this
+    is an API where each request authenticates itself, you want to return
+    ``nul``.
+
+**onAuthenticationFailure(Request $request, AuthenticationException $exception)**
+    This is called if authentication fails. Your job
+    is to return the :class:`Symfony\\Component\\HttpFoundation\\Response`
+    object that should be sent to the client. The ``$exception`` will tell you
+    *what* went wrong during authentication.
+
+**start**
+    This is called if the client accesses a URI/resource that requires authentication,
+    but no authentication details were sent (i.e. you returned ``null`` from
+    ``getCredentialsFromRequest()``). Your job is to return a
+    :class:`Symfony\\Component\\HttpFoundation\\Response` object that helps
+    the user authenticate (e.g. a 401 response that says "token is missing!").
+
+**supportsRememberMe**
```

Comment:
The rest of the methods are explaind in the general way, but with some comments for the current example. But this method is just explained for the example, so we lack the gernal purpose explanation.